### PR TITLE
feat: adds utility launch for debugging a single test file

### PR DESCRIPTION
**Pull request by mschnee**
Partially implements: #4300
See: #4406

This has been split from #4406 as a more manageable, smaller PR.

This update adds a developer convenience Launch task to allow debugging a single unit test file in Visual Studio Code.  

## Adds
- **`bin/mocha-current-file.js`** - this is a utility script that attempts to determine the dist output of the given test file, and passes it off to mocha in-process.
- **VSCode: Debug Current Test File** launch configuration.  This launch configuration allows you to debug the current unit test file currently focused.  It uses the above helper to determine the correct name and path of the compiled JavaScript file to run.

## Modifies
- The launch for "Run Mocha Tests" so that it also runs on Windows.

## Checklist
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

### CODE REVIEW

**Comment by mschnee** [comment]
There are a couple of instances (e.g. debugging `packages/metadata/src/__tests__/decorator-factory.unit.ts`) where breakpoints don't get picked up as expected.  This could be the `"autoAttachChildProcesses": true,` in the launch configuration.

File: .vscode/launch.json [hunk 1]
```diff
@@ -18,6 +18,22 @@
...
+      "name": "Debug Current Test File",
+      "program": "${workspaceRoot}/packages/build/node_modules/.bin/_mocha",
+      "cwd": "${workspaceRoot}",
+      "autoAttachChildProcesses": true,
+      "args": [
+        "--config",
+        "${workspaceRoot}/packages/build/config/.mocharc.json",
+        "-t",
+        "0",
+        "$(node ${workspaceRoot}/packages/build/bin/get-dist-file ${file})"
```
Comment (raymondfeng) [inline .vscode/launch.json]: I assume this won't work on Windows, right?

Comment (mschnee) [reply .vscode/launch.json]: This has been bugging me, so I made a few changes:
- "Run Mocha tests" now has the correct windows-specific configuration.
- replaced the "Debug Current Test File" subshell by writing a program that invokes the mocha cli directly.
- removed the `autoAttachChildProcess` setting.
- confirmed I could set a breakpoint in `packages/metadata/src/decorator-factory.ts` and vscode would stop on it when "Debug Current Test File" is run on `packages/metadata/src/__tests__/unit/decorator-factory.unit.ts`

**Comment by mschnee** [comment]
NOTE: I have rebased from upstream and squashed all commits, which should address the travis commitlint and doc errors.

File: packages/build/bin/mocha-current-file.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env node
```
Comment (raymondfeng) [inline packages/build/bin/mocha-current-file.js]: If this file is only used for `loopback-next` repo, please move it to `loopback-next/bin`.

Comment (mschnee) [reply packages/build/bin/mocha-current-file.js]: Done!

File: bin/mocha-current-file.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env node
+
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/build
```
Comment (raymondfeng) [inline bin/mocha-current-file.js]: Let's fix the headers.

File: bin/mocha-current-file.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        "0",
+        "${file}"
+      ],
+      "disableOptimisticBPs": true
+    }
+  ]
+}
+```
+
+For your personal projects, you can invoke directly from @loopback/build:
```
Comment (raymondfeng) [inline bin/mocha-current-file.js]: This is no longer true.

File: bin/mocha-current-file.js (hunk 3 above)
Comment (mschnee) [reply bin/mocha-current-file.js]: ...done?  It's not consistent in the files in this directory.

File: bin/mocha-current-file.js (hunk 4 above)
Comment (mschnee) [reply bin/mocha-current-file.js]: Removed.

File: .vscode/launch.json [hunk 5]
```diff
@@ -9,14 +9,34 @@
...
       "args": [
         "--config",
         "${workspaceRoot}/packages/build/config/.mocharc.json",
         "packages/*/dist/__tests__/**/*.js",
         "-t",
         "0"
-      ]
+      ],
+      "windows": {
+        "program": "${workspaceRoot}/packages/build/node_modules/mocha/bin/mocha",
```
Comment (bajtos) [inline .vscode/launch.json]: I think this part is no longer needed, see https://github.com/strongloop/loopback-next/pull/4643/commits/992dfe98c4c10087ed30125e5e4faed734f8a890

File: .vscode/launch.json [hunk 6]
```diff
@@ -9,14 +9,34 @@
       "request": "launch",
       "name": "Run Mocha tests",
       "program": "${workspaceRoot}/packages/build/node_modules/.bin/_mocha",
+      "runtimeArgs": ["-r", "${workspaceRoot}/packages/build/node_modules/source-map-support/register"],
```
Comment (bajtos) [inline .vscode/launch.json]: Nice 👍

File: .vscode/launch.json [hunk 7]
```diff
@@ -9,14 +9,34 @@
...
+      "cwd": "${workspaceRoot}",
+      "args": [
+        "--config",
+        "${workspaceRoot}/packages/build/config/.mocharc.json",
+        "-t",
+        "0",
+        "${file}"
+      ],
+      "disableOptimisticBPs": true,
```
Comment (bajtos) [inline .vscode/launch.json]: Can you please add a comment to explain why is this flag needed?

**Comment by mschnee** [comment]
Sorry for the delay in getting back to this PR!  I had a brief vacation.  Will update and address things :)

File: .vscode/launch.json (hunk 5 above)
Comment (mschnee) [reply .vscode/launch.json]: Done!

File: .vscode/launch.json (hunk 7 above)
Comment (mschnee) [reply .vscode/launch.json]: Done!

**Comment by mschnee** [comment]
With approvals from @raymondfeng and @bajtos I'd be happy to merge this in :)

File: bin/mocha-current-file.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env node
+
+// Copyright IBM Corp. 2018. All Rights Reserved.
```
Comment (raymondfeng) [inline bin/mocha-current-file.js]: 2018 -> 2020.

File: bin/mocha-current-file.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+```jsonc
+{
+  "version": "0.2.0",
+  "configurations": [
+    {
+      "type": "node",
+      "request": "launch",
+      "name": "Debug Current Test File",
+      "program": "${workspaceRoot}/bin/mocha-current-file",
+      "cwd": "${workspaceRoot}",
```
Comment (derdeka) [inline bin/mocha-current-file.js]: Should this block be identical to launch.json? runtimeArgs is missing.

File: .vscode/launch.json (hunk 5 above)
Comment (bajtos) [reply .vscode/launch.json]: I think we misunderstood each other. I was trying to say that the `windows` section is no longer needed, because the main `program` setting was already updated to work on all platforms (Windows, Linux, MacOS).

I _think_ that your current config will actually break debugging Mocha tests on Windows, because `mocha` is just a thin wrapper that spawns another child process (running `_mocha` script).

Anyhow, let's remove the `windows` section please, it shouldn't be needed anymore.

File: bin/mocha-current-file.js (hunk 9 above)
Comment (bajtos) [reply bin/mocha-current-file.js]: Yes please, let's include `runtimeArgs` too.

File: bin/mocha-current-file.js (hunk 8 above)
Comment (mschnee) [reply bin/mocha-current-file.js]: Done!

File: bin/mocha-current-file.js (hunk 9 above)
Comment (mschnee) [reply bin/mocha-current-file.js]: Added!

File: .vscode/launch.json (hunk 5 above)
Comment (mschnee) [reply .vscode/launch.json]: Fixed!

**Comment by mschnee** [comment]
Addressing the commitlint changes :)

**Comment by mschnee** [comment]
@derdeka What's the correct scope?  `core`?

**Comment by bajtos** [comment]
> What's the correct scope? core?

`core` is for changing affecting `@loopback/core` package.

In this case, I think it's best to leave the scope out. Also please note that we are using imperative form ("add something"), not present tense ("adds something"). See [subject](https://loopback.io/doc/en/lb4/code-contrib-lb4.html#subject) in our Commit Message Guidelines and also [A Note About Git Commit Messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).

I am proposing the following commit message:

```
feat: add a launch task for debugging a single test file
```

(Personally, I would choose `chore` instead of `feat` because this change is not visible to LoopBack users, but that does not really matter.)

File: bin/mocha-current-file.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+  let isPackageRoot = fs.existsSync(path.resolve(currentDir, 'package.json'));
+  while (!isPackageRoot) {
+    currentDir = path.join(currentDir, '..');
+    isPackageRoot = fs.existsSync(path.resolve(currentDir, 'package.json'));
+  }
+  const base = path.resolve(currentDir);
+  const relative = path.relative(currentDir, absolutePath);
+  const resultPath = relative.replace(/^src/, 'dist').replace(/\.ts$/, '.js');
+  return path.resolve(base, resultPath);
+}
```
Comment (raymondfeng) [inline bin/mocha-current-file.js]: Nitpick: a more robust implementation can be:
```js
// Find root directory of the file system
const root = path.parse(process.cwd()).root;

let pkgFile = path.resolve(filename, '../package.json');
while (!fs.existsSync(pkgFile)) {
  if (path.dirname(pkgFile) === root) {
    // We reach the root of file system
    pkgFile = undefined;
    break;
  }
  // Try the parent directory level package.json
  pkgFile = path.resolve(pkgFile, '../../package.json');
}

if (!pkgFile) return filename; // Or throw an error
const projectRoot = path.dirname(pkgFile);
const relative = path.relative(projectRoot, path.resolve(filename));
const jsFilePath = relative.replace(/^src/, 'dist').replace(/\.ts$/, '.js');
return path.resolve(projectRoot, jsFilePath);
```

Comment (mschnee) [reply bin/mocha-current-file.js]: I made a small change to the current function, and tested it by invoking this cli against both a test file inside of the `loopback-next` workspace, and also a `fake.test.js` file that has no `package.json` file in it's filesystem heirarchy

File: bin/mocha-current-file.js [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+function findDistFile(filename) {
+  const absolutePath = path.resolve(filename);
+  const systemRoot = path.parse(absolutePath).root;
+  let currentDir = path.dirname(absolutePath);
+
+  let isPackageRoot = fs.existsSync(path.resolve(currentDir, 'package.json'));
+  while (!isPackageRoot) {
+    currentDir = path.join(currentDir, '..');
+    if (path.dirname(currentDir) === systemRoot) {
```
Comment (raymondfeng) [inline bin/mocha-current-file.js]: This test should happen before change `currentDir` to its parent.

Comment (mschnee) [reply bin/mocha-current-file.js]: Good catch, updated :)

Comment (raymondfeng) [reply bin/mocha-current-file.js]: Nitpick again :-). I think you code will throw if `/myproject` is the current dir and no `package.json` is found there. It's really an edge case. So feel free to move forward.

Comment (mschnee) [reply bin/mocha-current-file.js]: Very edge case!  Since the idea is to run a single unit test inside one of the repo package directories, it's likely an expected error if you try to debug a file at the system room.
